teps/0003-tekton-oci-bundles.md

```diff
@@ -0,0 +1,258 @@
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This proposal is to be able to bundle Tasks (and Pipelines, and
+Resources, and potentially other future config objects) into an [OCI
+Artifact](https://github.com/opencontainers/artifacts), pushed to an
+image registry, and referenced from that registry.
+
+This is a TEP for the **Spec** of the Tekton OCI bundle that were
+discussed in the following docs:
+
+- [Tekton OCI Image Catalog](https://docs.google.com/document/d/1zUVrIbGZh2R9dawKQ9Hm1Cx3GevKIfOcRO3fFLdmBDc/edit#heading=h.tp9mko2koenr)
+- [Tekton OCI Image Design](https://docs.google.com/document/d/1lXF_SvLwl6OqqGy8JbpSXRj4hWJ6CSImlxlIl4V9rnM/edit?pl=1#)
+
+This will be based from the knowledge acquired on the [experimental oci
+project](https://github.com/tektoncd/experimental/tree/master/oci) we ran.
+
+## Motivation
+
+Today, TaskRuns can be defined completely (user specifies
+`.spec.taskSpec.steps` etc.), or by referencing a Task that must have
+previously been defined in the cluster's API server (user specifies
+`.spec.taskRef.name` and optionally `.namespace` -- if a Task doesn't
+exist with the specified name in the TaskRun's namespace, we check if
+it's a ClusterTask).
+
+This is overly limiting. It makes versioning hard, and it makes
+rolling out changes to Task definitions hard.
+
+We can leverage existing tooling and infrastructure used for sharing
+OCI images. An OCI image is really just a format for a set
+of binary files that can be uploaded to a registry (think DockerHub)
+with a tagged version and digest. It can easily be used to version and
+share Tasks and other Tekton resources.
+
+### Goals
+
+The goal of this TEP is to define the specification for the Tekton OCI
+bundles. The goal of the Tekton OCI bundles is to define a Spec on top
+of a known, widely used transport mechanisms, which are the OCI
+artifact (OCI images, …).
+
+### Non-Goals
+
+How Tekton OCI bundles are used (referred to) in the different Tekton
+components is out of scope and will be the subject for an additional
+TEP. Some examples in the doc might suggest how we would reference
+those in Pipeline CRDs but they are only examples, not propositions.
+
+The tooling around Tekton OCI bundles is also out of scope.
+
+
+## Requirements
+
+- Images have a defined format that allows Tasks, Pipelines, and any
+  other Tekton specific types to be written in, and read out.
+- The spec should be documented well enough such that any user could
+  write tooling to generate a valid Tekton Bundle
+
+## Proposal
+
+When using a Tekton Bundle in a task or pipeline reference, the OCI artifact backing the
+bundle must adhere to the following contract.
+
+Our most important capability is enabling references of remote Tasks
+from within TaskRuns or Pipelines and Pipelines from within
+PipelineRuns.
```

Comment:
I struggle to understand this. E.g. a new `PipelineRun` is created, but it contains a reference to a `Pipeline` declared within an image. So the controller - instead of accessing a resource through _informers_ and via ApiServer and etcd - it should now start to pull images from a registry - or multiple different registries - then extract data from the images (I imaging this is a kind of unzipping or similar) and then - no caching? - so next time the run is reconciled - the same procedure is happening? This is unclear and confusing for me - with the information that is provided so far.

Reply:
What we define here is merely the contract so the implementation details such as, is the image cached, are intentionally left out. But to your earlier sentiment, that is indeed correct. Rather than use the Kubernetes API (which is how Tekton looks up references at the moment), we would like to enable external references. This solves use cases of versioning, idempotency, etc.

Reply:
I need to update the doc with some user stories and usage example. As @pierretasci the scope of this TEP is *only* to define the *specification* of the OCI artifact that would "transport" the tekton definitions. We can add examples of how we think they could be used by the CLI, the pipeline controller, etc… but it's out of the scope here.
I'll add those examples next week :wink: